src/cbmc_starter_kit/template-for-ci-workflow/proof_ci.yaml

```diff
@@ -0,0 +1,246 @@
+# Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+# SPDX-License-Identifier: MIT-0
+name: Run CBMC proofs
+on:
+  push:
+    branches-ignore:
+      - gh-pages
+
+# USAGE
+#
+# By default, the most recently released and available version will be used
+# for CBMC, CBMC viewer, Litani and for the kissat, cadical SAT solvers.
+#
+# A semantic version can be specified for CBMC, CBMC viewer, Litani
+# (e.g. '5.70.0' for CBMC, '3.6' for CBMC viewer, '1.27.0' for Litani).
+# Known tags can be specified for kissat and cadical (e.g. 'rel-3.0.0' and
+# 'rel-1.5.3' respectively).
+#
+# GITHUB_ACTIONS_RUNNER is the name of the GitHub-hosted runner (e.g. 'ubuntu-20.04')
+# that will execute this workflow. THe runner must be on Ubuntu 20.04
+# If your repository has large runners at its disposal, then the value for this
+# key can be set to the name of your large runner. Find more details at:
+# https://docs.github.com/en/actions/using-github-hosted-runners/using-larger-runners
+
+
+env:
+  CADICAL_TAG: <__CADICAL_TAG__>
+  CBMC_VERSION: <__CBMC_VERSION__>
+  CBMC_VIEWER_VERSION: <__CBMC_VIEWER_VERSION__>
+  GITHUB_ACTIONS_RUNNER: <__GITHUB_ACTIONS_RUNNER__>
+  KISSAT_TAG: <__KISSAT_TAG__>
+  LITANI_VERSION: <__LITANI_VERSION__>
+  PROOFS_DIR: <__PROOFS_DIR__>
+  RUN_CBMC_PROOFS_COMMAND: ./run-cbmc-proofs.py
+
+
+jobs:
+  run_cbmc_proofs:
+    runs-on: <__GITHUB_ACTIONS_RUNNER__>
+    name: run_cbmc_proofs
+    permissions:
+      contents: read
+      id-token: write
+      pull-requests: read
+    steps:
+      - name: Check out repository and submodules recursively
+        uses: actions/checkout@v3
+        with:
+          submodules: 'recursive'
+      - name: Ensure CBMC, CBMC viewer, Litani versions have been specified
+        id: ensure_versions
+        if: ${{ env.CBMC_VERSION == '' || env.CBMC_VIEWER_VERSION == '' || env.LITANI_VERSION == '' }}
+        shell: bash
+        run: |
+          should_exit=false
+          if [ "${{ env.CBMC_VERSION }}" == "" ]; then
+            echo "You must specify a CBMC version (e.g. 'latest' or '5.70.0')"
+            should_exit=true
+          fi
+          if [ "${{ env.CBMC_VIEWER_VERSION }}" == "" ]; then
+            echo "You must specify a CBMC viewer version (e.g. 'latest' or '3.6')"
+            should_exit=true
+          fi
+          if [ "${{ env.LITANI_VERSION }}" == "" ]; then
+            echo "You must specify a Litani version (e.g. 'latest' or '1.27.0')"
+            should_exit=true
+          fi
+          if should_exit=true; then exit 1; fi
```

Comment:
I will be keeping this as is.

Reply:
The code isn't correct, it only works by accident. You could fix it by just writing `exit 1` at the end rather than having the conditional.

There's no need to set `should_exit` or have that variable at all: *the fact that we even enter this step means that something went wrong* (because of the `if: ${{ env.CBMC_VERSION == '' || ...` guard, so `should_exit` is vacuously true on all paths through this step. However, this is masked by the coding error on the final line, which assigns true to `should_exit` no matter what happened before.

Reply:
Again, have a look at the code I wrote at the top of this comment. The conditional isn't doing what you wanted it to do because it's assigning rather than evaluating. What is wrong here:
* To check for equality in shell you need double-brackets plus spaces around the `=` sign
```
# correct
if [[ $a == $b ]]
# wrong, this always evaluates to true
if a=b
```
* `should_exist` is a variable so it needs `$` when you read it (but not when you write it). So the correct way to write that test is
```
if [[ "$should_exist" == true ]];
```
But as I said above, I think the variable and test are actually both unnecessary

Reply:
> The code isn't correct, it only works by accident. 

I disagree with this statement.

> You could fix it by just writing exit 1 at the end rather than having the conditional.

If you approach it the way that you suggest, and both `CBMC_VERSION` and `LITANI_VERSION` happen to be empty, then you will exit too early. With my approach, you see a statement being printed out for each of the tools that happen to be empty.

Reply:
> so should_exit is vacuously true on all paths through this step.

If `should_exit` was always `true` then this step would always exit with return code of 1, which is not the case based on what we have seen in past executions.

You are right in that I can be using `if [[ "$should_exist" == true ]];`. I was using that equality check for the other if-statements above.